// astro/dist/content/types-generator.js
import glob from "fast-glob";
import { cyan } from "kleur/colors";
import * as path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";
import { normalizePath } from "vite";
import { AstroError, AstroErrorData } from "../core/errors/index.js";
import { info, warn } from "../core/logger/core.js";
import { isRelativePath } from "../core/path.js";
import { CONTENT_TYPES_FILE } from "./consts.js";
import {
  getContentEntryExts,
  getContentPaths,
  getEntryInfo,
  getEntrySlug,
  getEntryType,
  loadContentConfig,
  NoCollectionError,
  parseFrontmatter
} from "./utils.js";
class UnsupportedFileTypeError extends Error {
}
async function createContentTypesGenerator({
  contentConfigObserver,
  fs,
  logging,
  settings,
  viteServer
}) {
  const contentTypes = {};
  const contentPaths = getContentPaths(settings.config, fs);
  const contentEntryExts = getContentEntryExts(settings);
  let events = [];
  let debounceTimeout;
  const typeTemplateContent = await fs.promises.readFile(contentPaths.typesTemplate, "utf-8");
  async function init() {
    if (!fs.existsSync(contentPaths.contentDir)) {
      return { typesGenerated: false, reason: "no-content-dir" };
    }
    events.push({
      type: { name: "add", entry: contentPaths.config.url },
      opts: { logLevel: "warn" }
    });
    const globResult = await glob("**", {
      cwd: fileURLToPath(contentPaths.contentDir),
      fs: {
        readdir: fs.readdir.bind(fs),
        readdirSync: fs.readdirSync.bind(fs)
      }
    });
    const entries = globResult.map((e) => new URL(e, contentPaths.contentDir)).filter(
      // Config loading handled first. Avoid running twice.
      (e) => !e.href.startsWith(contentPaths.config.url.href)
    );
    for (const entry of entries) {
      events.push({ type: { name: "add", entry }, opts: { logLevel: "warn" } });
    }
    await runEvents();
    return { typesGenerated: true };
  }
  async function handleEvent(event, opts) {
    var _a, _b;
    const logLevel = (opts == null ? void 0 : opts.logLevel) ?? "info";
    if (event.name === "addDir" || event.name === "unlinkDir") {
      const collection2 = normalizePath(
        path.relative(fileURLToPath(contentPaths.contentDir), fileURLToPath(event.entry))
      );
      const isCollectionEvent = collection2.split("/").length === 1;
      if (!isCollectionEvent)
        return { shouldGenerateTypes: false };
      switch (event.name) {
        case "addDir":
          addCollection(contentTypes, JSON.stringify(collection2));
          if (logLevel === "info") {
            info(logging, "content", `${cyan(collection2)} collection added`);
          }
          break;
        case "unlinkDir":
          removeCollection(contentTypes, JSON.stringify(collection2));
          break;
      }
      return { shouldGenerateTypes: true };
    }
    const fileType = getEntryType(fileURLToPath(event.entry), contentPaths, contentEntryExts);
    if (fileType === "ignored") {
      return { shouldGenerateTypes: false };
    }
    if (fileType === "config") {
      contentConfigObserver.set({ status: "loading" });
      try {
        const config = await loadContentConfig({ fs, settings, viteServer });
        if (config) {
          contentConfigObserver.set({ status: "loaded", config });
        } else {
          contentConfigObserver.set({ status: "does-not-exist" });
        }
      } catch (e) {
        contentConfigObserver.set({
          status: "error",
          error: e instanceof Error ? e : new AstroError(AstroErrorData.UnknownContentCollectionError)
        });
      }
      return { shouldGenerateTypes: true };
    }
    if (fileType === "unsupported") {
      if (event.name === "unlink") {
        return { shouldGenerateTypes: false };
      }
      const entryInfo2 = getEntryInfo({
        entry: event.entry,
        contentDir: contentPaths.contentDir,
        // Skip invalid file check. We already know it’s invalid.
        allowFilesOutsideCollection: true
      });
      return {
        shouldGenerateTypes: false,
        error: new UnsupportedFileTypeError(entryInfo2.id)
      };
    }
    const entryInfo = getEntryInfo({
      entry: event.entry,
      contentDir: contentPaths.contentDir
    });
    if (entryInfo instanceof NoCollectionError) {
      if (["info", "warn"].includes(logLevel)) {
        warn(
          logging,
          "content",
          `${cyan(
            normalizePath(
              path.relative(fileURLToPath(contentPaths.contentDir), fileURLToPath(event.entry))
            )
          )} must be nested in a collection directory. Skipping.`
        );
      }
      return { shouldGenerateTypes: false };
    }
    const { id, collection } = entryInfo;
    const collectionKey = JSON.stringify(collection);
    const entryKey = JSON.stringify(id);
    switch (event.name) {
      case "add":
        const addedSlug = await parseSlug({ fs, event, entryInfo });
        if (!(collectionKey in contentTypes)) {
          addCollection(contentTypes, collectionKey);
        }
        if (!(entryKey in contentTypes[collectionKey])) {
          setEntry(contentTypes, collectionKey, entryKey, addedSlug);
        }
        return { shouldGenerateTypes: true };
      case "unlink":
        if (collectionKey in contentTypes && entryKey in contentTypes[collectionKey]) {
          removeEntry(contentTypes, collectionKey, entryKey);
        }
        return { shouldGenerateTypes: true };
      case "change":
        const changedSlug = await parseSlug({ fs, event, entryInfo });
        if (((_b = (_a = contentTypes[collectionKey]) == null ? void 0 : _a[entryKey]) == null ? void 0 : _b.slug) !== changedSlug) {
          setEntry(contentTypes, collectionKey, entryKey, changedSlug);
          return { shouldGenerateTypes: true };
        }
        return { shouldGenerateTypes: false };
    }
  }
  function queueEvent(rawEvent, opts) {
    const event = {
      type: {
        entry: pathToFileURL(rawEvent.entry),
        name: rawEvent.name
      },
      opts
    };
    if (!event.type.entry.pathname.startsWith(contentPaths.contentDir.pathname))
      return;
    events.push(event);
    debounceTimeout && clearTimeout(debounceTimeout);
    const runEventsSafe = async () => {
      try {
        await runEvents(opts);
      } catch {
      }
    };
    debounceTimeout = setTimeout(
      runEventsSafe,
      50
      /* debounce to batch chokidar events */
    );
  }
  async function runEvents(opts) {
    const logLevel = (opts == null ? void 0 : opts.logLevel) ?? "info";
    const eventResponses = [];
    for (const event of events) {
      const response = await handleEvent(event.type, event.opts);
      eventResponses.push(response);
    }
    events = [];
    let unsupportedFiles = [];
    for (const response of eventResponses) {
      if (response.error instanceof UnsupportedFileTypeError) {
        unsupportedFiles.push(response.error.message);
      }
    }
    if (unsupportedFiles.length > 0 && ["info", "warn"].includes(logLevel)) {
      warn(
        logging,
        "content",
        `Unsupported file types found. Prefix with an underscore (\`_\`) to ignore:
- ${unsupportedFiles.join(
          "\n"
        )}`
      );
    }
    const observable = contentConfigObserver.get();
    if (eventResponses.some((r) => r.shouldGenerateTypes)) {
      await writeContentFiles({
        fs,
        contentTypes,
        contentPaths,
        typeTemplateContent,
        contentConfig: observable.status === "loaded" ? observable.config : void 0,
        contentEntryTypes: settings.contentEntryTypes
      });
      if (observable.status === "loaded" && ["info", "warn"].includes(logLevel)) {
        warnNonexistentCollections({
          logging,
          contentConfig: observable.config,
          contentTypes
        });
      }
    }
  }
  return { init, queueEvent };
}
function addCollection(contentMap, collectionKey) {
  contentMap[collectionKey] = {};
}
function removeCollection(contentMap, collectionKey) {
  delete contentMap[collectionKey];
}
async function parseSlug({
  fs,
  event,
  entryInfo
}) {
  const rawContents = await fs.promises.readFile(event.entry, "utf-8");
  const { data: frontmatter } = parseFrontmatter(rawContents, fileURLToPath(event.entry));
  return getEntrySlug({ ...entryInfo, unvalidatedSlug: frontmatter.slug });
}
function setEntry(contentTypes, collectionKey, entryKey, slug) {
  contentTypes[collectionKey][entryKey] = { slug };
}
function removeEntry(contentTypes, collectionKey, entryKey) {
  delete contentTypes[collectionKey][entryKey];
}
async function writeContentFiles({
  fs,
  contentPaths,
  contentTypes,
  typeTemplateContent,
  contentEntryTypes,
  contentConfig
}) {
  let contentTypesStr = "";
  const collectionKeys = Object.keys(contentTypes).sort();
  for (const collectionKey of collectionKeys) {
    const collectionConfig = contentConfig == null ? void 0 : contentConfig.collections[JSON.parse(collectionKey)];
    contentTypesStr += `${collectionKey}: {
`;
    const entryKeys = Object.keys(contentTypes[collectionKey]).sort();
    for (const entryKey of entryKeys) {
      const entryMetadata = contentTypes[collectionKey][entryKey];
      const dataType = (collectionConfig == null ? void 0 : collectionConfig.schema) ? `InferEntrySchema<${collectionKey}>` : "any";
      const renderType = `{ render(): Render[${JSON.stringify(
        path.extname(JSON.parse(entryKey))
      )}] }`;
      const slugType = JSON.stringify(entryMetadata.slug);
      contentTypesStr += `${entryKey}: {
  id: ${entryKey},
  slug: ${slugType},
  body: string,
  collection: ${collectionKey},
  data: ${dataType}
} & ${renderType},
`;
    }
    contentTypesStr += `},
`;
  }
  if (!fs.existsSync(contentPaths.cacheDir)) {
    fs.mkdirSync(contentPaths.cacheDir, { recursive: true });
  }
  let configPathRelativeToCacheDir = normalizePath(
    path.relative(contentPaths.cacheDir.pathname, contentPaths.config.url.pathname)
  );
  if (!isRelativePath(configPathRelativeToCacheDir))
    configPathRelativeToCacheDir = "./" + configPathRelativeToCacheDir;
  if (configPathRelativeToCacheDir.endsWith(".ts")) {
    configPathRelativeToCacheDir = configPathRelativeToCacheDir.replace(/\.ts$/, "");
  }
  for (const contentEntryType of contentEntryTypes) {
    if (contentEntryType.contentModuleTypes) {
      typeTemplateContent = contentEntryType.contentModuleTypes + "\n" + typeTemplateContent;
    }
  }
  typeTemplateContent = typeTemplateContent.replace("// @@ENTRY_MAP@@", contentTypesStr);
  typeTemplateContent = typeTemplateContent.replace(
    "'@@CONTENT_CONFIG_TYPE@@'",
    contentConfig ? `typeof import(${JSON.stringify(configPathRelativeToCacheDir)})` : "never"
  );
  await fs.promises.writeFile(
    new URL(CONTENT_TYPES_FILE, contentPaths.cacheDir),
    typeTemplateContent
  );
}
function warnNonexistentCollections({
  contentConfig,
  contentTypes,
  logging
}) {
  for (const configuredCollection in contentConfig.collections) {
    if (!contentTypes[JSON.stringify(configuredCollection)]) {
      warn(
        logging,
        "content",
        `${JSON.stringify(
          configuredCollection
        )} is not a collection. Check your content config for typos.`
      );
    }
  }
}
export {
  createContentTypesGenerator
};
