// astro/dist/content/vite-plugin-content-imports.js
import * as devalue from "devalue";
import { extname } from "node:path";
import { pathToFileURL } from "url";
import { AstroErrorData } from "../core/errors/errors-data.js";
import { AstroError } from "../core/errors/errors.js";
import { escapeViteEnvReferences, getFileInfo } from "../vite-plugin-utils/index.js";
import { CONTENT_FLAG } from "./consts.js";
import {
  getContentEntryExts,
  getContentPaths,
  getEntryData,
  getEntryInfo,
  getEntrySlug,
  getEntryType,
  globalContentConfigObserver,
  NoCollectionError
} from "./utils.js";
function isContentFlagImport(viteId) {
  const flags = new URLSearchParams(viteId.split("?")[1]);
  return flags.has(CONTENT_FLAG);
}
function getContentRendererByViteId(viteId, settings) {
  let ext = viteId.split(".").pop();
  if (!ext)
    return void 0;
  for (const contentEntryType of settings.contentEntryTypes) {
    if (Boolean(contentEntryType.getRenderModule) && contentEntryType.extensions.includes("." + ext)) {
      return contentEntryType.getRenderModule;
    }
  }
  return void 0;
}
const CHOKIDAR_MODIFIED_EVENTS = ["add", "unlink", "change"];
function astroContentImportPlugin({
  fs,
  settings
}) {
  const contentPaths = getContentPaths(settings.config, fs);
  const contentEntryExts = getContentEntryExts(settings);
  const contentEntryExtToParser = /* @__PURE__ */ new Map();
  for (const entryType of settings.contentEntryTypes) {
    for (const ext of entryType.extensions) {
      contentEntryExtToParser.set(ext, entryType);
    }
  }
  const plugins = [
    {
      name: "astro:content-imports",
      async transform(_, viteId) {
        if (isContentFlagImport(viteId)) {
          const fileId = viteId.split("?")[0];
          const { id, slug, collection, body, data, _internal } = await setContentEntryModuleCache({
            fileId,
            pluginContext: this
          });
          const code = escapeViteEnvReferences(`
						export const id = ${JSON.stringify(id)};
						export const collection = ${JSON.stringify(collection)};
						export const slug = ${JSON.stringify(slug)};
						export const body = ${JSON.stringify(body)};
						export const data = ${devalue.uneval(data)};
						export const _internal = {
							filePath: ${JSON.stringify(_internal.filePath)},
							rawData: ${JSON.stringify(_internal.rawData)},
						};`);
          return { code, map: { mappings: "" } };
        }
      },
      configureServer(viteServer) {
        viteServer.watcher.on("all", async (event, entry) => {
          if (CHOKIDAR_MODIFIED_EVENTS.includes(event) && getEntryType(entry, contentPaths, contentEntryExts) === "config") {
            for (const modUrl of viteServer.moduleGraph.urlToModuleMap.keys()) {
              if (isContentFlagImport(modUrl) || Boolean(getContentRendererByViteId(modUrl, settings))) {
                const mod = await viteServer.moduleGraph.getModuleByUrl(modUrl);
                if (mod) {
                  viteServer.moduleGraph.invalidateModule(mod);
                }
              }
            }
          }
        });
      }
    }
  ];
  if (settings.contentEntryTypes.some((t) => t.getRenderModule)) {
    plugins.push({
      name: "astro:content-render-imports",
      async load(viteId) {
        const contentRenderer = getContentRendererByViteId(viteId, settings);
        if (!contentRenderer)
          return;
        const { fileId } = getFileInfo(viteId, settings.config);
        const entry = await getContentEntryModuleFromCache(fileId);
        if (!entry) {
          throw new AstroError({
            ...AstroErrorData.UnknownContentCollectionError,
            message: `Unable to render ${JSON.stringify(
              fileId
            )}. Did you import this module directly without using a content collection query?`
          });
        }
        return contentRenderer.bind(this)({ entry, viteId });
      }
    });
  }
  const contentEntryModuleByIdCache = /* @__PURE__ */ new Map();
  function isAwaitingQueue(cacheEntry) {
    return typeof cacheEntry === "object" && cacheEntry != null && "awaitingQueue" in cacheEntry;
  }
  function getContentEntryModuleFromCache(id) {
    const cacheEntry = contentEntryModuleByIdCache.get(id);
    if (isAwaitingQueue(cacheEntry)) {
      return new Promise((resolve, reject) => {
        cacheEntry.awaitingQueue.push(resolve);
      });
    } else if (cacheEntry) {
      return Promise.resolve(cacheEntry);
    }
    return Promise.resolve(void 0);
  }
  async function setContentEntryModuleCache({
    fileId,
    pluginContext
  }) {
    contentEntryModuleByIdCache.set(fileId, { awaitingQueue: [] });
    const contentConfig = await getContentConfigFromGlobal();
    const rawContents = await fs.promises.readFile(fileId, "utf-8");
    const fileExt = extname(fileId);
    if (!contentEntryExtToParser.has(fileExt)) {
      throw new AstroError({
        ...AstroErrorData.UnknownContentCollectionError,
        message: `No parser found for content entry ${JSON.stringify(
          fileId
        )}. Did you apply an integration for this file type?`
      });
    }
    const contentEntryParser = contentEntryExtToParser.get(fileExt);
    const {
      rawData,
      body,
      slug: unvalidatedSlug,
      data: unvalidatedData
    } = await contentEntryParser.getEntryInfo({
      fileUrl: pathToFileURL(fileId),
      contents: rawContents
    });
    const entryInfoResult = getEntryInfo({
      entry: pathToFileURL(fileId),
      contentDir: contentPaths.contentDir
    });
    if (entryInfoResult instanceof NoCollectionError)
      throw entryInfoResult;
    const { id, slug: generatedSlug, collection } = entryInfoResult;
    const _internal = { filePath: fileId, rawData };
    const slug = getEntrySlug({ id, collection, slug: generatedSlug, unvalidatedSlug });
    const collectionConfig = contentConfig == null ? void 0 : contentConfig.collections[collection];
    let data = collectionConfig ? await getEntryData(
      { id, collection, slug, _internal, unvalidatedData },
      collectionConfig,
      pluginContext,
      settings
    ) : unvalidatedData;
    const contentEntryModule = {
      id,
      slug,
      collection,
      data,
      body,
      _internal
    };
    const cacheEntry = contentEntryModuleByIdCache.get(fileId);
    if (isAwaitingQueue(cacheEntry)) {
      for (const resolve of cacheEntry.awaitingQueue) {
        resolve(contentEntryModule);
      }
    }
    contentEntryModuleByIdCache.set(fileId, contentEntryModule);
    return contentEntryModule;
  }
  return plugins;
}
async function getContentConfigFromGlobal() {
  const observable = globalContentConfigObserver.get();
  if (observable.status === "init") {
    throw new AstroError({
      ...AstroErrorData.UnknownContentCollectionError,
      message: "Content config failed to load."
    });
  }
  if (observable.status === "error") {
    throw observable.error;
  }
  let contentConfig = observable.status === "loaded" ? observable.config : void 0;
  if (observable.status === "loading") {
    contentConfig = await new Promise((resolve) => {
      const unsubscribe = globalContentConfigObserver.subscribe((ctx) => {
        if (ctx.status === "loaded") {
          resolve(ctx.config);
          unsubscribe();
        }
        if (ctx.status === "error") {
          resolve(void 0);
          unsubscribe();
        }
      });
    });
  }
  return contentConfig;
}
export {
  astroContentImportPlugin
};
