// astro/dist/assets/vendor/image-size/index.js
import * as fs from "fs";
import * as path from "path";
import Queue from "../queue/queue.js";
import { detector } from "./detector.js";
import { typeHandlers } from "./types.js";
const MaxBufferSize = 512 * 1024;
const queue = new Queue({ concurrency: 100, autostart: true });
const globalOptions = {
  disabledFS: false,
  disabledTypes: []
};
function lookup(buffer, filepath) {
  const type = detector(buffer);
  if (typeof type !== "undefined") {
    if (globalOptions.disabledTypes.indexOf(type) > -1) {
      throw new TypeError("disabled file type: " + type);
    }
    if (type in typeHandlers) {
      const size = typeHandlers[type].calculate(buffer, filepath);
      if (size !== void 0) {
        size.type = type;
        return size;
      }
    }
  }
  throw new TypeError(
    "unsupported file type: " + type + " (file: " + filepath + ")"
  );
}
async function asyncFileToBuffer(filepath) {
  const handle = await fs.promises.open(filepath, "r");
  const { size } = await handle.stat();
  if (size <= 0) {
    await handle.close();
    throw new Error("Empty file");
  }
  const bufferSize = Math.min(size, MaxBufferSize);
  const buffer = Buffer.alloc(bufferSize);
  await handle.read(buffer, 0, bufferSize, 0);
  await handle.close();
  return buffer;
}
function syncFileToBuffer(filepath) {
  const descriptor = fs.openSync(filepath, "r");
  const { size } = fs.fstatSync(descriptor);
  if (size <= 0) {
    fs.closeSync(descriptor);
    throw new Error("Empty file");
  }
  const bufferSize = Math.min(size, MaxBufferSize);
  const buffer = Buffer.alloc(bufferSize);
  fs.readSync(descriptor, buffer, 0, bufferSize, 0);
  fs.closeSync(descriptor);
  return buffer;
}
var image_size_default = imageSize;
function imageSize(input, callback) {
  if (Buffer.isBuffer(input)) {
    return lookup(input);
  }
  if (typeof input !== "string" || globalOptions.disabledFS) {
    throw new TypeError("invalid invocation. input should be a Buffer");
  }
  const filepath = path.resolve(input);
  if (typeof callback === "function") {
    queue.push(
      () => asyncFileToBuffer(filepath).then(
        (buffer) => process.nextTick(callback, null, lookup(buffer, filepath))
      ).catch(callback)
    );
  } else {
    const buffer = syncFileToBuffer(filepath);
    return lookup(buffer, filepath);
  }
}
const disableFS = (v) => {
  globalOptions.disabledFS = v;
};
const disableTypes = (types2) => {
  globalOptions.disabledTypes = types2;
};
const setConcurrency = (c) => {
  queue.concurrency = c;
};
const types = Object.keys(typeHandlers);
export {
  image_size_default as default,
  disableFS,
  disableTypes,
  imageSize,
  setConcurrency,
  types
};
