// astro/dist/core/config/timer.js
import fs from "fs";
class AstroTimer {
  constructor() {
    this.ongoingTimers = /* @__PURE__ */ new Map();
    this.stats = {};
    this.enabled = !!process.env.ASTRO_TIMER_PATH;
  }
  /**
   * Start a timer for a scope with a given name.
   */
  start(name) {
    var _a;
    if (!this.enabled)
      return;
    (_a = globalThis.gc) == null ? void 0 : _a.call(globalThis);
    this.ongoingTimers.set(name, {
      startTime: performance.now(),
      startHeap: process.memoryUsage().heapUsed
    });
  }
  /**
   * End a timer for a scope with a given name.
   */
  end(name) {
    var _a;
    if (!this.enabled)
      return;
    const stat = this.ongoingTimers.get(name);
    if (!stat)
      return;
    (_a = globalThis.gc) == null ? void 0 : _a.call(globalThis);
    const endHeap = process.memoryUsage().heapUsed;
    this.stats[name] = {
      elapsedTime: performance.now() - stat.startTime,
      heapUsedChange: endHeap - stat.startHeap,
      heapUsedTotal: endHeap
    };
    this.ongoingTimers.delete(name);
  }
  /**
   * Write stats to `process.env.ASTRO_TIMER_PATH`
   */
  writeStats() {
    if (!this.enabled)
      return;
    fs.writeFileSync(process.env.ASTRO_TIMER_PATH, JSON.stringify(this.stats, null, 2));
  }
}
export {
  AstroTimer
};
