// astro/dist/assets/vendor/image-size/types/tiff.js
import * as fs from "fs";
import { readUInt } from "../readUInt.js";
function readIFD(buffer, filepath, isBigEndian) {
  const ifdOffset = readUInt(buffer, 32, 4, isBigEndian);
  let bufferSize = 1024;
  const fileSize = fs.statSync(filepath).size;
  if (ifdOffset + bufferSize > fileSize) {
    bufferSize = fileSize - ifdOffset - 10;
  }
  const endBuffer = Buffer.alloc(bufferSize);
  const descriptor = fs.openSync(filepath, "r");
  fs.readSync(descriptor, endBuffer, 0, bufferSize, ifdOffset);
  fs.closeSync(descriptor);
  return endBuffer.slice(2);
}
function readValue(buffer, isBigEndian) {
  const low = readUInt(buffer, 16, 8, isBigEndian);
  const high = readUInt(buffer, 16, 10, isBigEndian);
  return (high << 16) + low;
}
function nextTag(buffer) {
  if (buffer.length > 24) {
    return buffer.slice(12);
  }
}
function extractTags(buffer, isBigEndian) {
  const tags = {};
  let temp = buffer;
  while (temp && temp.length) {
    const code = readUInt(temp, 16, 0, isBigEndian);
    const type = readUInt(temp, 16, 2, isBigEndian);
    const length = readUInt(temp, 32, 4, isBigEndian);
    if (code === 0) {
      break;
    } else {
      if (length === 1 && (type === 3 || type === 4)) {
        tags[code] = readValue(temp, isBigEndian);
      }
      temp = nextTag(temp);
    }
  }
  return tags;
}
function determineEndianness(buffer) {
  const signature = buffer.toString("ascii", 0, 2);
  if ("II" === signature) {
    return "LE";
  } else if ("MM" === signature) {
    return "BE";
  }
}
const signatures = [
  // '492049', // currently not supported
  "49492a00",
  // Little endian
  "4d4d002a"
  // Big Endian
  // '4d4d002a', // BigTIFF > 4GB. currently not supported
];
const TIFF = {
  validate(buffer) {
    return signatures.includes(buffer.toString("hex", 0, 4));
  },
  calculate(buffer, filepath) {
    if (!filepath) {
      throw new TypeError("Tiff doesn't support buffer");
    }
    const isBigEndian = determineEndianness(buffer) === "BE";
    const ifdBuffer = readIFD(buffer, filepath, isBigEndian);
    const tags = extractTags(ifdBuffer, isBigEndian);
    const width = tags[256];
    const height = tags[257];
    if (!width || !height) {
      throw new TypeError("Invalid Tiff. Missing tags");
    }
    return { height, width };
  }
};
export {
  TIFF
};
